### DISCUSSION TITLE
Get snaphot while ringing

### DISCUSSION THREAD

**Discussion by Tiffytaff** [dgreif/ring#604] [Q&A] [answered]
Hallo,
I hope this is the right place to ask my question.
I am a newbie in JavaScript. However i want to get a snapshot when someone press the ring button. In this case I want to send the snapshot with an Express server.
I take one  of your samples and use the onNewDing event.
But in this event I can't get a snapshot. Maybe the reason is that I can't use await in a sync function or maybe the ring doorbell can't send a snapshot during video recording.

Here is what I did so far

```
import 'dotenv/config'
import { RingApi } from 'ring-client-api'

const express = require('express');
const app = express();
const port = 3000;


async function example() {
  const { env } = process,
    ringApi = new RingApi({
      // Replace with your refresh token
      refreshToken: 'xxx',
      locationIds: ['123'],
      cameraDingsPollingSeconds: 2
    }),
    locations = await ringApi.getLocations(),
    location = locations[0],
    cameras = await ringApi.getCameras(),
    camera =   cameras[0]

    pict = await camera.getSnapshot()

    app.get('/file.jpg', (req, res) => {
    res.contentType('image/jpeg');
    res.send(pict);   // this works, but is not an actual snapshot
  });


camera.onNewDing.subscribe((ding) => {
    const event =
      ding.kind === 'motion'
        ? 'Motion detected'
        : ding.kind === 'ding'
        ? 'Doorbell pressed'
        : `Video started (${ding.kind})`
        
        // How can I get an actual snapshot here?
       
    console.log(
      `${event} on ${camera.name} camera. Ding id ${
        ding.id_str
      }.  Received at ${new Date()}`
    )

  })
```

Can someone help me?
Thanks

**Comment by dgreif** [comment, marked as answer]
This is a great place to ask 😄.  First, it's important to know which type of camera you are using. If it is a camera equipped with a battery, you _won't_ be able to pull a new snapshot while recording is in progress.  See https://github.com/dgreif/ring/wiki/Snapshot-Limitations for details.

Beyond that limitation, it's hard to say what the exact problem is.  `pict` will be a `Buffer`, and you should be able to send that as a response with express, but I don't work with express enough to know the right way to do that.  Google should be able to provide a decent answer on that one 😉 

Regarding getting the snapshot in the `onNewDing` callback, you can add `async` to the callback like `camera.onNewDing.subscribe(async (ding) => {`, which will enable you to do `await` inside it. How you update the endpoint to serve the newly captured snapshot is up to you.

**Reply by Tiffytaff** [reply]
Thanks for quick reaction,
My hardware is a Ring Doorbell 3 plus.
I added the async to the callback and it works fine. Sending pict as response with express works also.
So far I'm happy.

If I press the ring button then the recording starts and I can not get a snapshot. I understand this is a limitation of the ring doorbell.
Your recommendation in https://github.com/dgreif/ring/wiki/Snapshot-Limitations is to use the live view instead. How can I access it?
My attempt:

```
camera.onNewDing.subscribe(async(ding) => {
    const event =
      ding.kind === 'motion'
        ? 'Motion detected'
        : ding.kind === 'ding'
        ? 'Doorbell pressed'
        : `Video started (${ding.kind})`
        

        console.log(ding.id_str);
        var recUrl = await camera.getRecordingUrl(
          ding.id_str
        );
        console.log(recUrl);
       
    console.log(
      `${event} on ${camera.name} camera. Ding id ${
        ding.id_str
      }.  Received at ${new Date()}`
    )
```


This ends up with:
```
6949099408143945431
(node:12952) UnhandledPromiseRejectionWarning: Error: Not found with response: 
    at RingRestClient.<anonymous> (c:\Test06\ringapp4\node_modules\ring-client-api\lib\api\rest-client.js:229:27)
    at Generator.throw (<anonymous>)
    at rejected (c:\Test06\ringapp4\node_modules\ring-client-api\lib\api\rest-client.js:6:65)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:12952) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:12952) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```
Another thing that I notice:
On my mobile I get a notification with a snapshot from the ring app, when I pressed the doorbell button. This snapshot would be perfect, but I guess it is not possible to get it with the api. Isn't it?

**Reply by dgreif** [reply]
Here is an example from another project that grabs a snapshot from live stream: https://github.com/tsightler/ring-mqtt/blob/dde93f9a7ffaad92fea34af8f085f070fd3170a4/devices/camera.js#L516-L548. If I recall correctly, this has the negative side effect of ending the recording on the Ring servers after about 1 second, so keep that in mind.  

Unfortunately I don't currently have a way to pull the image from those mobile notifications, but i agree those would be ideal.  I've been trying to work directly with Ring to get access to those, but no luck so far.

**Reply by Tiffytaff** [reply]
I checked it. tsightler any you did a very good job. Thanks for that.
To solve my problem, I will try to create a little android app that get the notification image
and provides it with a http-server.
